### PR TITLE
10.0 Add sale_project_user_profile

### PR DESCRIPTION
**Pull request by leemannd**
Original work is from @jcoux .
This module allow to define a mapping for project user task.
The goal is to sell (at customer) services with different user profiles (Project manager, developer).
The mapping allow to the project manager to decide which user profile is used in timesheet invoicing for given task.

@sergio-teruel @gurneyalex Told me that this feature could be of some interest for you. While I'm updating to OCA standards, you can already have a look at it.

### TODO
- [x] Update to fit OCA standards
- [x] Not fully tested yet

### CODE REVIEW

**Comment by leemannd** [comment]
@rafaelbn If you have the time to look at it :)

**Comment by rafaelbn** [comment]
Hi @leemannd ! Thanks for this contribution! 😄 

Please review files changed! It looks you are trying to modify in this PR files from other modules! 😮 

Ping me when reviewed and runbot green (now is red) 

Thanks!

**Comment by leemannd** [comment]
@rafaelbn Travis is green now, it was a wrong rebase. I just forced the rebuild of the runbot.

**Comment by rafaelbn** [comment]
hi @leemannd @gurneyalex I'm trying to test it in runbot but it doesn't work.... Are you able to test it?

See:

**Comment by leemannd** [comment]
@rafaelbn It's green now

**Comment by rafaelbn** [comment]
WoW! 😄 this works like a charm!

Just one more thing @leemannd . I tested in runbot but I could find in runbot the option of this module: https://github.com/OCA/sale-workflow/tree/10.0/sale_order_invoicing_finished_task

I would like to test them together to check alll is fine. Is it possible?

In any case 👍

**Review by rafaelbn — APPROVED** [review]
Tested in runbot! Please @chienandalu @luismontalba test this one. Thanks!

File: sale_project_user_profile/models/account_analytic_line.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            task = self.task_id
+            sale_line_map = task.project_user_task_sale_line_map_ids.filtered(
+                lambda m: m.user_id == self.user_id
+            )
+
+            # 2- If we don't have mapping for the task,
+            # we search if we have mapping for user, without tasks
+            if not sale_line_map:
+                project = self.project_id
+                if project.project_uses_task_sale_line_map:
```
Comment (carlosdauden) [inline sale_project_user_profile/models/account_analytic_line.py]: Checked above:
https://github.com/OCA/sale-workflow/pull/537/files#diff-bf6374bc95e90f0a694756f62373f688R17

File: sale_project_user_profile/models/account_analytic_line.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                    )
+
+            # 3- If we don't found mapping for the user,
+            # we raise a validation error
+            if not sale_line_map:
+                raise ValidationError(
+                    _(
+                        'You cannot timesheet on this task. '
+                        'Contact the project manager %s'
+                    ) % self.project_id.user_id.name
```
Comment (carlosdauden) [inline sale_project_user_profile/models/account_analytic_line.py]: Use project
https://github.com/OCA/sale-workflow/pull/537/files#diff-bf6374bc95e90f0a694756f62373f688R30
or remove when remove line 31

File: sale_project_user_profile/models/project_user_task_sale_line_map.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+from odoo.exceptions import ValidationError
+
+
+class ProjectUserTaskSalelineMap(models.Model):
+    _name = 'project.user.task.sale_line.map'
+
+    @api.constrains('project_id', 'user_id', 'task_id',)
+    def _check_unique_project_user_task(self):
+        domain = [
+            ('project_id', '=', self.project_id.id or False),
```
Comment (carlosdauden) [inline sale_project_user_profile/models/project_user_task_sale_line_map.py:15]: `or False` unnecessary

File: sale_project_user_profile/models/project_user_task_sale_line_map.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        ]
+        if self.search(domain):
+            raise ValidationError(_(
+                'A project user task mapping must be unique '
+                'by project, user and task!'
+            ))
+
+    name = fields.Char(
+        compute='_compute_name',
+        readonly=True,
```
Comment (carlosdauden) [inline sale_project_user_profile/models/project_user_task_sale_line_map.py:28]: Unnecessary.
`inverse` parameter is needed to allow set value in compute field

**Review by luismontalba — APPROVED** [review]
Tested on runbot, working fine

File: sale_project_user_profile/README.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+=========================
+Sale Project User Profile
+=========================
+
+This module allow to define a mapping for project user task.
+The goal is to sell (at customer) services with different user profiles (Project manager, developer).
+The mapping allow to the project manager to decide which user profile is used in timesheet invoicing for given task.
+
+
+=====
```
Comment (chienandalu) [inline sale_project_user_profile/README.rst]: drop this line so it's h3

File: sale_project_user_profile/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    'website': 'http://www.camptocamp.com',
+    'depends': [
+        'hr_timesheet',
+        'product',
+        'project',
+        'sale',
+        'sale_timesheet',
+    ],
+    'data': [
+        # Security
```
Comment (chienandalu) [inline sale_project_user_profile/__manifest__.py]: I think it self explains :wink:

File: sale_project_user_profile/i18n/sale_project_user_profile.pot [hunk 7]
```diff
@@ -0,0 +1,142 @@
+# Translation of Odoo Server.
```
Comment (chienandalu) [inline sale_project_user_profile/i18n/sale_project_user_profile.pot]: this file is not needed

File: sale_project_user_profile/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    'name': 'Sale Project User Profile',
+    'version': '10.0.1.0.0',
+    'author': 'Camptocamp SA,Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'category': 'Project',
+    'website': 'http://www.camptocamp.com',
```
Comment (chienandalu) [inline sale_project_user_profile/__manifest__.py]: Follow new OCA guidelines regarding website in manifest: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#modules

File: sale_project_user_profile/models/account_analytic_line.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+from odoo import models, _
+from odoo.exceptions import ValidationError
+
+
+class AccountAnalyticLine(models.Model):
+    _inherit = 'account.analytic.line'
+
+    def _get_sale_order_line(self, vals=None):
+        new_vals = dict(vals or {})
+
```
Comment (chienandalu) [inline sale_project_user_profile/models/account_analytic_line.py]: No need to add extra blank lines inside the method.

File: sale_project_user_profile/models/product_template.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+# Copyright 2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    track_service = fields.Selection(
```
Comment (carlosdauden) [inline sale_project_user_profile/models/product_template.py:11]: IMHO I think you should use a new boolean field to maintain compatibility.
If you want I can make a PR to your branch

If this module is installed on any client you can add a script:
`UPDATE new_boolean_field = true, track_service = 'timesheet' WHERE track_service='user_profile'`

**Review by carlosdauden — CHANGES_REQUESTED** [review]

File: sale_project_user_profile/models/sale_order.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+                        "With a service product of type 'User profile', "
+                        "all service products must have 'User profile' type."
+                    ))
+
+    @api.multi
+    def action_confirm(self):
+        # Copy of the same implementation in sale_timesheet module,
+        # but with track_service = 'user_profile'
+        result = super(SaleOrder, self).action_confirm()
+        for order in self:
```
Comment (chienandalu) [inline sale_project_user_profile/models/sale_order.py:36]: Maybe you could reduce nesting with something of the sort and avoid the loop break also:
```python
for order in self:
    if order.project_project_id:
        continue
    if order.order_line.filtered(
        lambda x: x.track_service == 'user_profile')[0]:
```

File: sale_project_user_profile/models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    @api.constrains(
+        'state',
+        'order_line',
+    )
+    def _check_state(self):
```
Comment (chienandalu) [inline sale_project_user_profile/models/sale_order.py:16]: Put a blank line before

**Review by chienandalu — CHANGES_REQUESTED** [review]
Thanks for your work :slightly_smiling_face: Some code comments

**Comment by carlosdauden** [comment]
@leemannd I'm working in it :wink:

File: sale_project_user_profile/__manifest__.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    'depends': [
+        'hr_timesheet',
+        'product',
+        'project',
+        'sale',
```
Comment (bealdav) [inline sale_project_user_profile/__manifest__.py:15]: Thanks for your contribution. It seems there is redundancy in depends.

**Comment by carlosdauden** [comment]
https://github.com/leemannd/sale-workflow/pull/9

**Comment by carlosdauden** [comment]
@leemannd Changes done here:
https://github.com/leemannd/sale-workflow/pull/9

If this module is installed in any client you can change version and add migration script with:
`UPDATE new_boolean_field = true, track_service = 'timesheet' WHERE track_service='user_profile'`

File: sale_project_user_profile/models/sale_order.py (hunk 12 above)
Comment (carlosdauden) [reply sale_project_user_profile/models/sale_order.py:16]: Is a decorator

File: sale_project_user_profile/models/sale_order.py (hunk 11 above)
Comment (carlosdauden) [reply sale_project_user_profile/models/sale_order.py:36]: Use of filtered is less efficient because for break with first match

**Comment by rafaelbn** [comment]
Hi @leemannd @gurneyalex ,

We propose changes in https://github.com/leemannd/sale-workflow/pull/9 because if not this module could be incompatible with other modules. In fact it is in this moment incompatible with sale_order_invoicing_finished_task 

Trying to help! 😄

**Comment by leemannd** [comment]
@rafaelbn @carlosdauden @chienandalu I've made small changes given feedbacks. Sorry for the delay.
I'm putting me a reminder in order to try to finish it next week.

**Comment by rafaelbn** [comment]
Please @leemannd could you check why travis is red? Thanks a lot

**Review by rafaelbn — COMMENTED** [review]
Travis says:

```
2017-12-05 18:36:46,592 5750 INFO openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: test_04_project_user_task_sale_line_mapping (odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile.TestSaleProjectUserProfile)

2017-12-05 18:36:50,156 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: FAIL

2017-12-05 18:36:50,161 5750 INFO openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: ======================================================================

2017-12-05 18:36:50,161 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: FAIL: test_04_project_user_task_sale_line_mapping (odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile.TestSaleProjectUserProfile)

2017-12-05 18:36:50,161 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: Traceback (most recent call last):

2017-12-05 18:36:50,162 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: `   File "/home/travis/build/OCA/sale-workflow/sale_project_user_profile/tests/test_sale_project_user_profile.py", line 239, in test_04_project_user_task_sale_line_mapping

2017-12-05 18:36:50,162 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: `     self.create_timesheet_line(self.user_1, project, task_1, 1)

2017-12-05 18:36:50,162 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: `   File "/usr/lib/python2.7/contextlib.py", line 24, in __exit__

2017-12-05 18:36:50,162 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: `     self.gen.next()

2017-12-05 18:36:50,162 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: `   File "/home/travis/odoo-10.0/odoo/tests/common.py", line 122, in _assertRaises

2017-12-05 18:36:50,163 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: `     yield cm

2017-12-05 18:36:50,163 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: ` AssertionError: ValidationError not raised

2017-12-05 18:36:50,163 5750 INFO openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: Ran 4 tests in 20.470s

2017-12-05 18:36:50,164 5750 ERROR openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile: FAILED

2017-12-05 18:36:50,164 5750 INFO openerp_test odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile:  (failures=1)

2017-12-05 18:36:50,164 5750 INFO openerp_test odoo.modules.module: odoo.addons.sale_project_user_profile.tests.test_sale_project_user_profile tested in 20.47s, 20531 queries

2017-12-05 18:36:50,164 5750 ERROR openerp_test odoo.modules.module: Module sale_project_user_profile: 1 failures, 0 errors

2017-12-05 18:36:50,457 5750 INFO openerp_test odoo.modules.registry: module sale_automatic_workflow: creating or updating database tables

2017-12-05 18:36:50,850 5750 INFO openerp_test odoo.models: Storing computed values of sale.order fields all_qty_delivered
```

**Comment by pedrobaeza** [comment]
Please include here our PR that solves also Travis problems.

File: sale_project_user_profile/README.rst [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+Usage
+=====
+
+* You can define a new type of tracking service on product template: 'User profile'.
+* When you confirm a sale order with a service of this type, a project is automatically created.
+* On the project, you can define a mapping for user, task and sale order line.
+* When you define a timesheet on the project, the sale order line used in account analytic line is get in the mapping.
+
+
```
Comment (JayVora-SerpentCS) [inline sale_project_user_profile/README.rst:22]: missing Runbot writeup

**Review by JayVora-SerpentCS — CHANGES_REQUESTED** [review]

**Comment by leemannd** [comment]
Closing as it

**Comment by rafaelbn** [comment]
I reopen to test functionality with @alexey-pelykh

**Comment by alexey-pelykh** [comment]
@rafaelbn thanks for pointing this module out, I wish I'd seen it earlier - there's a #765 for v12 that addresses very similar thing based on [project roles](https://odoo-community.org/shop/product/project-roles-4565) and its [support of timesheets](https://odoo-community.org/shop/product/task-logs-by-role-4568). Maybe @leemannd and others would be interested in reviewing those as well.

**Comment by rafaelbn** [comment]
Closing again. Thanks

**Pull request closed without merging**